### ISSUE TITLE
FR: Type Hints/Stubs

### ISSUE THREAD

**Issue by adam-grant-hendry** [pyqtgraph/pyqtgraph#2409]
### Short description

Feature request: Could you please add type hints to the source (and/or in stub files) and add a `py.typed` marker file per [PEP 561](https://peps.python.org/pep-0561/)

### Code to reproduce

N/A

### Tested environment(s)

N/A

### Additional context

This is a feature request.

**Comment by j9ac9k** [comment]
Hi @adam-grant-hendry 

We've had _lots_ of discussions about this behind the scenes.  This is a major undertaking, but we do feel like it is worthwhile now that per NEP-29 we mandate NumPy 1.20+ so we have access to `numpy.typing.ArrayLike`.  

That said, we would love some help here from contributors with how to roll something like this out on a partial basis.  We won't be able to roll out something across the library in one go, so we'll have to do it bit by bit.   This blog post on how the urllib3 devs did it I found to be of interest: https://sethmlarson.dev/blog/tests-arent-enough-case-study-after-adding-types-to-urllib3

**Comment by adam-grant-hendry** [comment]
@j9ac9k I've been helping the `pyvista` and `pyvistaqt` teams get up to speed on their type hinting. Gradual typing is very possible; simply add `partial\n` within your shipped `py.typed` file.

I use `PySide6` mostly and find the [IceSpringPySideStubs](https://github.com/baijifeilong/IceSpringPySideStubs) useful in this regard. For the other python `Qt` bindings, [python-qt-tools](https://github.com/python-qt-tools) has stubs for `PyQt5`, `PySide2`, and `PyQt6`.

Since `pyqtgraph` is heavily `Qt` based, specifically with the Graphics View framework, I think it should be relatively straightforward matter to begin gradual typing PRs.

Per my experience,  would recommend the following:

1. Choose a static typing package to adhere to. I recommend `mypy`, but `pyright` (now included within `pylint`) catches many useful things as well. Not all static typers are created equal (`pyright` is Microsoft, `pyre` is Facebook, `mypy` the de-fact Python checker, `pytype` is Google). I've only ever used `mypy` and `pylint`/`pyright`.

2. Add a `py.typed` marker with `partial/n` in the file to indicate the package is partially typed

3. Add `mypy` to your pre-commit hooks, which will help with your users submitting PRs.

After this, you can tackle one module at a time. It's usually easiest/best to start with the most used module in the package and work forward from there.

**Comment by adam-grant-hendry** [comment]
@j9ac9k If you like my approach, I can create a PR to add a `py.typed` marker to the package and put `mypy` into the pre-commit hooks. That's the first step.

Since it looks like (at first glance) nothing is type hinted as of yet, `mypy` will pick up all modules as dynamically typed (i.e. not statically typed at all), which should mean `pre-commit` wouldn't generate a huge amount of errors (if any).

**Comment by j9ac9k** [comment]
Hi @adam-grant-hendry 

I meant to reply right away but I've had my head stuck in the sand trying to get as far in PR 1753 as I could 😆 

Regarding the first point regarding supporting the variety of bindings; I think mirroring the setup that qtpy has with respect to type-hinting would be ideal.  We currently have a [stub file](https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/Qt/__init__.pyi) to get type-hints in PyCharm, but it doesn't work w/ pyright, so it's just a band-aid.  Unlike qtpy, we try and conform to a Qt6 API more than Qt5, but I think that difference can be settled w/o too much difficulty.

I too have primarily used mypy, but I use pyright when working within vscode which I do sometimes.  I'm not opposed to using multiple type-checking tools provided the pain point isn't too bad 😆 

Once we have mypy runs and have it be able to read partial annotations without issue, adding a pre-commit hook will definitely happen.

Hadn't heard of `py.typed` before, but that does seem to be a necessary feature at a glance!

I think efforts on this should begin after we release 0.13.0; which _should_ happen before much longer (I'm trying to sneak in a few other existing PRs first).

Any effort in this area, especially setup in such a way that we can _gradually_ roll out type-annotations that can be read in by a variety of editors will have widespread support among maintainers here.

**Comment by j9ac9k** [comment]
there are a few functions that I am scratching my head with what would be a "good" way to annotate them in such a way it's readable; [mkColor](https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/functions.py#L232) is one of those functions that accepts a wide variety of types and every time I try and annotate it, it just looks like a mess.  If you have an idea on how to annotate it in a readable fashion, please share!

**Comment by NeilGirdhar** [comment]
> [mkColor](https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/functions.py#L232) is one of those functions that accepts a wide variety of types and every time I try and annotate it, it just looks like a mess. If you have an idea on how to annotate it in a readable fashion, please share!

You just use overloads:
```python
from typing import overload

@overload
def mkColor(color: str, /): ...
@overload
def mkColor(r: int, g: int, b: int, a: int = 255, /): ...
@overload
def mkColor(rgb: tuple[int, int, int], /): ...
@overload
def mkColor(rgba: tuple[int, int, int, int], /): ...
@overload
def mkColor(greyscale: float, /): ...
@overload
def mkColor(int_hues: tuple[int, int], /): ...

def mkColor(*args):
    ...
```

**Comment by rwarren** [comment]
Is there any progress/update here?  Or any starting work that I can pick up?

**Comment by j9ac9k** [comment]
> Is there any progress/update here? Or any starting work that I can pick up?

We have an outstanding PR with type annotations for the Flowchart module, now that I got 0.13.5 released (and subsequent regressions addressed), that was next on my queue.

I would pick a smaller module, and add type annotations there.  I will verify if they're correct (probably using pyright instead of mypy, as more editors support pyright checks inside the editor) before merging though.

I highly recommend you start off with a small module, or a smaller GraphicsItem class.

**Comment by j9ac9k** [comment]
> > [mkColor](https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/functions.py#L232) is one of those functions that accepts a wide variety of types and every time I try and annotate it, it just looks like a mess. If you have an idea on how to annotate it in a readable fashion, please share!
> 
> You just use overloads:
> 
> ```python
> from typing import overload
> 
> @overload
> def mkColor(color: str, /): ...
> @overload
> def mkColor(r: int, g: int, b: int, a: int = 255, /): ...
> @overload
> def mkColor(rgb: tuple[int, int, int], /): ...
> @overload
> def mkColor(rgba: tuple[int, int, int, int], /): ...
> @overload
> def mkColor(greyscale: float, /): ...
> @overload
> def mkColor(int_hues: tuple[int, int], /): ...
> 
> def mkColor(*args):
>     ...
> ```

While this works well for documenting `mkColor`, a _lot_ of methods in the library accept as arguments the same arguments that `mkColor` accepts, so we should almost certainly use an alias that can be imported and used elsewhere.

**Comment by NeilGirdhar** [comment]
@j9ac9k That's true.  You may want to tighten up the broadness of the interface in a few places?  One common design principle is there should not be too many ways of doing the same thing.

**Comment by j9ac9k** [comment]
> @j9ac9k That's true. You may want to tighten up the broadness of the interface in a few places? One common design principle is there should not be too many ways of doing the same thing.

Hey, first off, sorry to make your excellent suggest sit w/o an answer for over a year.  It's been a busy year for me so my pyqtgraph work was pretty minimal. 

I'm fine with narrowing down the type annotations a bit, but I need to be careful here as this is part of the public API so if I want to remove existing functionality I need to go through the crummy deprecation phase, and when I deprecate part of the public API, I try and have a good reason for it, and I'm not sure "make the type annotation cleaner" would make the cut.

That said, if we can make `mkColor` pass type checkers with a slimmer definition, I'd be down for using those definitions elsewhere, and effectively not advertise/document some of the less supported input args.

**Comment by NeilGirdhar** [comment]
Yeah, that makes perfect sense.  And no worries about the response.

Consider something like `mkPen`:
```python
def mkPen(*args, **kargs):
    """
    Convenience function for constructing QPen. 
    
    Examples::
    
        mkPen(color)
        mkPen(color, width=2)
        mkPen(cosmetic=False, width=4.5, color='r')
        mkPen({'color': "#FF0", width: 2})
        mkPen(None)   # (no pen)
    
    In these examples, *color* may be replaced with any arguments accepted by :func:`mkColor() <pyqtgraph.mkColor>`    """
    color = kargs.get('color', None)
    width = kargs.get('width', 1)
    style = kargs.get('style', None)
    dash = kargs.get('dash', None)
    cosmetic = kargs.get('cosmetic', True)
    hsv = kargs.get('hsv', None)
    
    if len(args) == 1:
        arg = args[0]
        if isinstance(arg, dict):
            return mkPen(**arg)
        if isinstance(arg, QtGui.QPen):
            return QtGui.QPen(arg)  ## return a copy of this pen
        elif arg is None:
            style = QtCore.Qt.PenStyle.NoPen
        else:
            color = arg
    if len(args) > 1:
        color = args
        
    if color is None:
        color = mkColor('l')
    if hsv is not None:
        color = hsvColor(*hsv)
    else:
        color = mkColor(color)
    pen = QtGui.QPen(QtGui.QBrush(color), width)
    pen.setCosmetic(cosmetic)
    if style is not None:
        pen.setStyle(style)
    if dash is not None:
        pen.setDashPattern(dash)

    # for width > 1.0, we are drawing many short segments to emulate a
    # single polyline. the default SquareCap style causes artifacts.
    # these artifacts can be avoided by using RoundCap.
    # this does have a performance penalty, so enable it only
    # for thicker line widths where the artifacts are visible.
    if width > 4.0:
        pen.setCapStyle(QtCore.Qt.PenCapStyle.RoundCap)

    return pen
```
A simpler (and less flexible) way of writing something like this would be:
```python
def mkPen(pen: Pen | None = None,
          *,
          width: int = 1,
          style: QtCore.Qt.PenStyle | None = None,
          dash: DashPattern | None = None,
          cosmetic: Cosmetic | None = None,
          color: Color | None = None,
          hsv: HSV | None = None) -> Pen:
    """Convenience function for constructing QPen. 
    
    Examples::
    
        mkPen(color=color)
        mkPen(color=color, width=2)
        mkPen(cosmetic=False, width=4.5, color='r')
        mkPen(color="#FF0", width=2)
        mkPen(style=QtCore.Qt.PenStyle.NoPen)   # (no pen)
    
    In these examples, *color* may be replaced with any arguments accepted by :func:`mkColor() <pyqtgraph.mkColor>`    """
    if color is None and hsv is None:
        color = mkColor('l')
    elif hsv is not None:
        color = hsvColor(*hsv)
    else:
        color = mkColor(color)
        
    pen = QtGui.QPen(QtGui.QBrush(color), width)
    pen.setCosmetic(cosmetic)
    if style is not None:
        pen.setStyle(style)
    if dash is not None:
        pen.setDashPattern(dash)

    # for width > 1.0, we are drawing many short segments to emulate a
    # single polyline. the default SquareCap style causes artifacts.
    # these artifacts can be avoided by using RoundCap.
    # this does have a performance penalty, so enable it only
    # for thicker line widths where the artifacts are visible.
    if width > 4.0:
        pen.setCapStyle(QtCore.Qt.PenCapStyle.RoundCap)

    return pen
```
I see your point about deprecation periods, so this may not be worth it for you.  This is just an illustration of how one might write these functions if we were starting from scratch and wanted type annotations.  I also feel like mentioning that I find the latter interface easier to understand (in fact the docstring isn't even really necessary).

**Comment by j9ac9k** [comment]
We could deprecate `mkPen()` outright, and create `makePen()` instead, and shift usage of the library to `makePen()`, given a long enough deprecation cycle, no end user code should break in the near or medium term...

I do like your second instance of having specific args, instead of outright accepting `*args`.  This makes things so much easier to define...

**Comment by rwarren** [comment]
Is the plan to do direct in-code typing, like is done in PR #2867?  Or to use `.pyi` files so various overloads can be set up?  A hybrid?

`mkPen` is a great example since it shows how the use of blanket `*args, **kwargs` function signatures makes it a bit tricky to create type hints directly in code.

To keep current compatibility (and avoid deprecation hell) it seems like `.pyi` stub files are probably the best way to go, no?  Then `@overloads` can be created for all of the cases they are needed, as mentioned by @NeilGirdhar above.

Also - I tired generating stubs with mypy's `stubgen` and got a decent start... but with a LOT of things to fill in. Has anybody tried this yet?  I also tried [pytype](https://github.com/google/pytype) in hopes that it could infer types better. but could not get it to work.

As an example, here is what `stubgen` automatically generates for `Canvas.pyi`:

```python
from ..Qt import QtWidgets
from ..graphicsItems.ROI import ROI
from _typeshed import Incomplete

__all__ = ['Canvas']

class Canvas(QtWidgets.QWidget):
    sigSelectionChanged: Incomplete
    sigItemTransformChanged: Incomplete
    sigItemTransformChangeFinished: Incomplete
    ui: Incomplete
    view: Incomplete
    itemList: Incomplete
    allowTransforms: Incomplete
    multiSelectBox: Incomplete
    redirect: Incomplete
    items: Incomplete
    grid: Incomplete
    hideBtn: Incomplete
    ctrlSize: int
    sizeApplied: bool
    registeredName: Incomplete
    menu: Incomplete
    def __init__(self, parent: Incomplete | None = None, allowTransforms: bool = True, hideCtrl: bool = False, name: Incomplete | None = None) -> None: ...
    def splitterMoved(self) -> None: ...
    def hideBtnClicked(self) -> None: ...
    def autoRange(self) -> None: ...
    def resizeEvent(self, ev: Incomplete | None = None) -> None: ...
    def updateRedirect(self, *args) -> None: ...
    def redirectItems(self, canvas) -> None: ...
    def reclaimItems(self) -> None: ...
    def treeItemChanged(self, item, col) -> None: ...
    def treeItemSelected(self) -> None: ...
    def selectedItems(self): ...
    def selectItem(self, item) -> None: ...
    def showMultiSelectBox(self) -> None: ...
    def mirrorSelectionClicked(self) -> None: ...
    def reflectSelectionClicked(self) -> None: ...
    def resetTransformsClicked(self) -> None: ...
    def multiSelectBoxChanged(self) -> None: ...
    def multiSelectBoxChangeFinished(self) -> None: ...
    def multiSelectBoxMoved(self) -> None: ...
    def addGraphicsItem(self, item, **opts): ...
    def addGroup(self, name, **kargs): ...
    def addItem(self, citem): ...
    def treeItemMoved(self, item, parent, index) -> None: ...
    def itemVisibilityChanged(self, item) -> None: ...
    def removeItem(self, item) -> None: ...
    def clear(self) -> None: ...
    def addToScene(self, item) -> None: ...
    def removeFromScene(self, item) -> None: ...
    def listItems(self): ...
    def getListItem(self, name): ...
    def itemTransformChanged(self, item) -> None: ...
    def itemTransformChangeFinished(self, item) -> None: ...
    menuItem: Incomplete
    def itemListContextMenuEvent(self, ev) -> None: ...
    def removeClicked(self) -> None: ...

class SelectBox(ROI):
    def __init__(self, scalable: bool = False) -> None: ...
```

**Comment by NeilGirdhar** [comment]
>  Or to use .pyi files so various overloads can be set up

I'm not sure what you mean?  You can set up overloads right in the code.  `pyi` is a good solution when maintainers are reluctant to add type annotations.  They don't have any special capabilities.

**Comment by rwarren** [comment]
> I'm not sure what you mean? You can set up overloads right in the code.

Didn't know you could do this!

> Typically, it's less work to write the annotations in the code, and it's easier to keep them synchronized as the code changes.

Agreed this is is nicer.  The nice part about `stubgen` is it typically gets a good chunk of the types for you automatically.  I've seen heard that you can get it to merge into source instead of creating `.pyi` files, but I've not tried it.

> Personally, I would start with PyRight since I think it's currently the most advanced type checker (in terms of what's been implemented). Of course, that's just a personal opinion.

With `pytype` I mean using it to infer types and generate `.pyi` files... which I thought was a feature, but is not (explaining why I could not get it to work).

fwiw, PyRight/PyLance are how I use typing daily... and occasionally mypy mixed in to see what else it picks up.

I'm going to type a few modules and see what it looks like. I'll post a diff here for you guys to check out.

**Comment by NeilGirdhar** [comment]
> I'm going to type a few modules and see what it looks like. I'll post a diff here for you guys to check out.

Awesome!  Just FYI in case this helps to mention: pyqtgraph is marked as requiring Python >= 3.10, which means you can use all [the new typing features from 3.10](https://docs.python.org/3/whatsnew/3.10.html).

**Comment by rwarren** [comment]
> Awesome! Just FYI in case this helps to mention: pyqtgraph is marked as requiring Python >= 3.10, which means you can use all [the new typing features from 3.10](https://docs.python.org/3/whatsnew/3.10.html).

I had already gone looking for this and was happy to see 3.10+.  I think this means the only inaccessible typing feature is 3.11's `Self` for factory classmethods... which I'm not sure even exist in the pyqtgraph codebase.

**Comment by rwarren** [comment]
While trying to determine the best entry point for getting this done I had a deeper look at PR #2867 and noticed a somewhat pedantic style issue... commented here:

https://github.com/pyqtgraph/pyqtgraph/pull/2867/files#r1590437201

In short: instance attributes should be typed at class level, as defined [here](https://mypy.readthedocs.io/en/stable/class_basics.html#instance-and-class-attributes) (and shown in my linked comment).

What I'm currently doing is deciding whether I start my typing based off that flowchart-focused PR, or just do the typing based off of `master`.  Any comments?

**Comment by NeilGirdhar** [comment]
> In short: instance attributes should be typed at class level, as defined [here](https://mypy.readthedocs.io/en/stable/class_basics.html#instance-and-class-attributes) (and shown in my linked comment).

Why?  That means you'll annotate them twice instead of once.

**Comment by NeilGirdhar** [comment]
> [](https://github.com/pyqtgraph/pyqtgraph/pull/2867/files#r1590437201)
> 
> In short: instance attributes should be typed at class level, as defined [here](https://mypy.readthedocs.io/en/stable/class_basics.html#instance-and-class-attributes) (and shown in my linked comment).

In your linked documentation, they suggest annotating in `__init__`, which is what I usually see?  I guess it doesn't matter, as long as you don't annotate things like:
```
class X:
  # Do not add this:
  # x: int

  def __init__(self, x: int):
    self.x = x  # Don't annotate this
```

Also, `Optional` is considered to be outdated by most of the typing people.  There's no reason to use it anymore.

**Comment by rwarren** [comment]
[PEP526](https://peps.python.org/pep-0526/#class-and-instance-variable-annotations) is what defines the "official" recommended typing for class and instance variables... but I'm not sure how official it really is.  I definitely thought it was the defined approach.  All kinds of combos work, as you say.  Any typing is better than no typing! :)

FWIW, I initially didn't like defining types for instance variables at the class level, but [typing.ClassVar](https://docs.python.org/3/library/typing.html#typing.ClassVar) is there to distinguish types for class properties vs instance properties.  I've been doing it the PEP526 way (instance var types at class level) for years now and am pretty happy with it and thought it was the acknowledged standard?  Curiously, how to type instance variables has not made it into the [typing best practices](https://typing.readthedocs.io/en/latest/source/best_practices.html) yet, so... 🤔 

And... since I was curious...

To see what level of adoption this class-level typing approach has, I went on a hunt through some popular python packages to see what others are doing (aside from standard lib compliance with things like dataclasses and `typing.NamedTuple`)... and it was a bit of an interesting mix (unsurprising due to the state of public libraries from a typing perspective)...

* Arrow -- follows PEP526, per [here](https://github.com/arrow-py/arrow/blob/master/arrow/parser.py#L158
* pyparsing -- follows PEP526, per [here](https://github.com/pyparsing/pyparsing/blob/master/pyparsing/results.py#L21)
* SQLAlchemy -- follows PEP526, per [here](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/pool/base.py#L61)
* aiohttp -- does NOT follow PEP526... types `__init__` args only per [here](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/client.py#L217)
* fastapi -- does NOT follow PEP526, and also types `__init__` args only, in an interestingly verbose way (per [here](https://github.com/tiangolo/fastapi/blob/master/fastapi/applications.py#L64))
* ... going to stop there ...

Whatever works, works!  Just trying to establish what the standard should be for pyqtgraph typing implementation.  I personally think that the PEP526 way is a really clean way to do it, since it also crisply defines what instance variables (and ClassVar variables) are available, with their types, in a single location.

**Comment by j9ac9k** [comment]
Love to see a discussion springing up about this!  I'll have more to say later (I have to pack for a work trip), but figure I'll chime in with my current thoughts.

*. I prefer inline annotations for a few reasons
  1. I don't have to put type annotations in the docstring, numpydoc (which I know isn't used everywhere but goal is for the library to use it everywhere) will make use of inline annotations
  2. (Warning, I could be quite mistaken here) I don't think stub files can be used to verify correctness _within_ the library itself, they are intended to be used by people using the library and verify their own code conforms to the stubs.
*. Per NEP-29, pyqtgraph is now Python 3.10+ (unless someone points out a major regression from the 0.13.5-0.13.7 releases in which case I'd address that while preserving the python 3.9+ dependency, but I'm getting more confident by the day that this isn't an issue).
  *. This upcoming python 3.10 minimum version was one of the reasons I didn't merge the flow chart PR, because I knew we could do away w/ typing extensions before long
*. For the specific type checker, I'm leaning towards using pyright as opposed to mypy, primarily because pyright has broader support in editors (primarily due to the language server). If it's all the same, having the same lint checker that the editors use would be preferable.

The other wrinkle here is the library, while not _huge_ is fairly large.  Trying to roll out type annotations across the board is just not feasible in 1 PR.  Whatever the mechanism that is rolled out, it needs to be incremental, and ideally the CI would be checking for whats already been added and shouldn't be generating too much noise for the bits that aren't annotated.


Lastly, I want to point out while I do type annotation on my own code, I haven't really rolled it out in larger projects before, so I could _really_ use guidance from the broader community (hence why I love seeing the discussion here).

**Comment by j9ac9k** [comment]
> While trying to determine the best entry point for getting this done I had a deeper look at PR #2867 and noticed a somewhat pedantic style issue... commented here:
> 
> https://github.com/pyqtgraph/pyqtgraph/pull/2867/files#r1590437201
> 
> In short: instance attributes should be typed at class level, as defined [here](https://mypy.readthedocs.io/en/stable/class_basics.html#instance-and-class-attributes) (and shown in my linked comment).
> 
> What I'm currently doing is deciding whether I start my typing based off that flowchart-focused PR, or just do the typing based off of `master`. Any comments?

The FlowChart PR was submitted by another contributor, and I really haven't given the PR the attention it deserves largely from my own limited time on the library combined with other elements of the library I deemed more important.  I would not view the "style" of the type-annotations in that PR to be a reflection of what the maintainers would like across the code-base.

**Comment by j9ac9k** [comment]
> > Awesome! Just FYI in case this helps to mention: pyqtgraph is marked as requiring Python >= 3.10, which means you can use all [the new typing features from 3.10](https://docs.python.org/3/whatsnew/3.10.html).
> 
> I had already gone looking for this and was happy to see 3.10+. I think this means the only inaccessible typing feature is 3.11's `Self` for factory classmethods... which I'm not sure even exists in the pyqtgraph codebase.

FWIW the library will adopt Python 3.11 in the first release after April 4th, 2025.  While it feels forever away, for the few _potential_ relevant bits of the codebase where the use of `Self` would be beneficial, I would just ignore them until pyqtgraph is python 3.11+.

**Comment by j9ac9k** [comment]
Oh, one last thing before I put the computer away for the night... the library _should_ make use of type aliases, `mkColor` comes to mind, which to fully type annotate all the acceptable types for that function, it would be _quite_ extensive.  We use the args there is a ton of places (same goes for `mkPen` and `mkBrush`).  Being able to document aliases (I sort of started a `color_like` alias in the [documentation here](https://github.com/pyqtgraph/pyqtgraph/blob/3d9831a119430f5fbcf7fa8ed11e846b68500a2e/doc/source/conf.py#L71) (but it just links to `mkColor`, it doesn't actually specify types.

**Comment by NeilGirdhar** [comment]
I don't have much to add except that I've never seen this kind of enthusiasm for typing and I just love it.  Also, agree with all the comments/decisions.

**Comment by j9ac9k** [comment]
Thanks for pointing out PEP526, I generally try and follow standards unless I have reason not to, and I can't think of a great reason not to follow it.

**Comment by rwarren** [comment]
After sleeping on it, and doing some exploratory work on options yesterday, my current thinking is this:

1. Any typing is better than no typing, but...
2. Only having types for part of the library is not great (but still better than no typing)
3. It is better to have poor (but available) typing for the whole library, with pockets of awesomeness (i.e. complete, or almost complete, typing)

To that end I propose this, and will work on a sample PR to review progress on in the next few days:

1. Long term objective: Full typing directly within every `.py` file in the `pyqtgraph` code base
    * Using PEP526 styling, as discussed above
2. Short term objective: Provide `.pyi` coverage for _all_ `.py` files in the code base
    * Coverage here can be partial, and can be auto-generated
        * auto-generated type files will contain more loose usage of `Any`, `Incomplete`, and empty container (e.g. just `dict`, `list`, and `tuple` with no member type info) than is ideal... but it's a start
    * this will result in a `.pyi` file adjacent to every `.py` file in  deployment, but this is typical (e.g. see all of numpy like here)
    * I will do this in a single commit as a part of the first demo PR 
        * Contrary to what was stated earlier, I do think this part can be done in a single PR (it has zero effect on runtime)
 3. Mid-term (while transitioning from short to long term): Replace `.pyi` files one-by-one completely by moving types into the matching `.py` file
     * `.pyi` files typically take precedence over typing in the `.py` file, so you must do one or the other
     * I will do a handful of `.py` files as a part of the first demo PR
 
Other notes...

### Regarding what type checker to use and/or comply with

Fundamentally: Type hinting is type hinting!  There is no typing for pyright vs mypy vs pytype vs pylizer vs whatever.  These are just the tools to check the typing!

That said... they all might report slightly differently, so there needs to be a standard used _within_ the project for zeroing out type warnings.  Pyright is a good way to go, but I suggest we also run mypy on top of this.  Both in strict mode (`--strict`) to ensure that things like missing return types and function arg types are not allowed.

### vscode project

Visual Studio Code (vscode) will be used by a huge number of people, and it makes doing type checking easier because there are extensions for both Pyright/PyLance and mypy.  When doing the typing, I'll certainly be using vscode to help get the job done.

I see no vscode project info in the pyqtgraph code base.  It might be useful to set up a basic one as part of this typing, just for internal pyqtgraph development.  Is there any objection to this?  Impact would be the addition of a `.vscode` dir at the root of the project, with a few json files in it.  This would be convenient... but not required.  pyright and mypy compliance testing can be done at the cli level as well.

----

I'm now going a bit far here, but I had thoughts/notes on related topics so I'll keep going, although obviously stretching a bit far for "just" getting some typing into pyqtgraph, but... on my mind for the typing so I might as well. :)

### Pre-commit hooks...

I see that there are some pre-commit hooks in place, which is great... it might be worth adding a pylyzer static type check as a pre-commit hook once typing is in place.  I say pylizer because it is FAST, as pre-commit hooks should be.

### PyQt5 vs PyQt6 vs PySide2 vs PySide6

With pyqtgraph supporting all of these, I'm a bit concerned this might be the biggest annoyance in the typing effort!  Qt wrapping typing is not perfect by any stretch (e.g. I still juggle between the official stubs, `IceSpringPySideStubs-PySide6`, and `PySide6-stubs` and have issues with each).  I expect this will be interesting... but am hoping that the fact that pyqtgraph wraps so many classes will make this less of a problem.  Also curious what you all find here with regards to core Qt typing... although I suspect there are few of us wanting typing to be comprehensive for all libraries, including Qt wrappers...

**Comment by NeilGirdhar** [comment]
Looking at the pyi file you illustrated above, I don't think the `pyi` generation is going to save you any time.  By default, all type checkers make better assumptions than what you illustrated.  I would just skip the `pyi`.  It's just more work and there's no benefit to it.

**Comment by rwarren** [comment]
> Also, citing Numpy is not a good example.

fwiw I only showed numpy as an example of how `.pyi` sit beside the `.py` files.  Definitely an entirely different rationale behind adding `.pyi`!!

> Looking at the pyi file you illustrated above, I don't think the pyi generation is going to save you any time. By default, all type checkers make better assumptions than what you illustrated. I would just skip the pyi. It's just more work and there's no benefit to it.

I'm more optimistic about this than that.  Please don't use the crappy stubgen-generated example I gave as what an auto-generated `.pyi` file will cover.  It *should* be much better than you expect... I hope.  I'm going to try generating the starter `.pyi` files using a tool (like DropBox's [pyannotate](https://github.com/dropbox/pyannotate)) that runs all unit tests and introspects types for everything.  That said, I've only barely fiddled with tools like that, so we'll see how it goes. :)  I've added typing to various medium sized projects 10k -50k lines), but never used the automatic tools.  Fingers crossed.

> If you decide to annotate the implementation too, that may be useful to the project development, but it has no effect on users.

I generally just type everything since it pays off at all layers.  I'll see if just doing the API helps... but that implies identifying the API clearly, so we'll see how it goes.

**Comment by j9ac9k** [comment]
Lots of good stuff here, but my main concern with stub files is, can we verify the library itself is conforming to the specified types of the stubs? Last time I looked at stub files, they were only used to verify things _using_ pyqtgraph were conforming to the annotations.

To add to the extra bit of complexity, pyqtgraph uses relative imports, which is also something I'm not sure stub files support/recognize.

**Comment by NeilGirdhar** [comment]
> I hope. I'm going to try generating the starter `.pyi` files using a tool (like DropBox's [pyannotate](https://github.com/dropbox/pyannotate)) that runs all unit tests and introspects types for everything. That said, I've only barely fiddled with tools like that, so we'll see how it goes. :) I've added typing to various medium sized projects 10k -50k lines), but never used the automatic tools. Fingers crossed.

Don't you think it's going to add to reviewer time too?  When your change is reviewed, the reviewer has to keep your change in one pane and the source code in another and verify that everything matches.   If it's inline, it's so much easier for the reviewer, and you as well.

Plus, you have a plan to eventually merge the `pyi` back into the code to make the annotations inline.  Someone has to review that giant commit!  And someone has to do it.  Don't  you want to save that whole process.

And I'm not entirely sure why you don't want to do it inline.  Is there some benefit?

> I've added typing to various medium sized projects 10k -50k lines), but never used the automatic tools. Fingers crossed.

I think if you're going to use the automatic tools, then you should still make your commit inline and do the entire merging in of the annotations right from the start.  Otherwise, you're creating a lot of complexity for other people?

> I generally just type everything since it pays off at all layers. I'll see if just doing the API helps... but that implies identifying the API clearly, so we'll see how it goes.

Totally agree with you there as long as the maintainer is okay with that.  I can see how annotating the implementation acts as an API test (since the implementation calls the API).

**Comment by bersbersbers** [comment]
> can we verify the library itself is conforming to the specified types of the stubs? Last time I looked at stub files, they were only used to verify things using pyqtgraph were conforming to the annotations.

You may check out `typeguard` (or its alteratives). It has a pytest plugin which means that you can runtime-verify types during testing without impacting testing runtime behavior.

**Comment by j9ac9k** [comment]
> > can we verify the library itself is conforming to the specified types of the stubs? Last time I looked at stub files, they were only used to verify things using pyqtgraph were conforming to the annotations.
> 
> You may check out `typeguard` (or its alteratives). It has a pytest plugin which means that you can runtime-verify types during testing without impacting non-testing runtime behavior.

Thanks for the suggestion w/ `typeguard`, hadn't heard of it before.  I think in our case, limiting ourselves to the code called from the test suites is probably not sufficient, our test coverage isn't ...great

**Comment by bersbersbers** [comment]
Not a problem at all, you can use the import hook and run whatever you like to run.

**Comment by j9ac9k** [comment]
Hey folks, had some travel, sorry I let the discussion die down.  I wanted to experiment how I wanted the documentation to look/render, you can see my work here: https://github.com/pyqtgraph/pyqtgraph/pull/3028

While working on this, I realized there is an incompatibility between numpy-docstring style, and `sphinx-autodoc-typehints`.  Upon further reflection, I decided that this might be ok, meaning it's ok if the inline annotations don't render in the docs directly.  While I think the inline annotations/stubs should be accurate/complete, the docstring should be a subset that convey intent instead of exhaustive type syntax that may be tougher to read.

**Comment by j9ac9k** [comment]
I also just came across numpydantic, looks like it may be useful for our use case, as there are many places we care about the numpy array shape and dtype.

**Comment by rwarren** [comment]
After many distractions I'm back at this with a lot of progress.  I have a formatting question...

Consider this method (on `matplotlibWidget`) as currently defined:

```python
    @typing.overload
    def __init__(self, figsize=(5.0, 4.0), dpi=100, parent=None):
        pass
```

How do you guys want to format this?

**OPTION 1:  Indented single line parameters**

Proper typing (with my preferred one-line-per-parameter and indentation/spacing) looks like this:

```python
    @typing.overload
    def __init__(self,
                 figsize: tuple[float, float] = (5.0, 4.0),
                 dpi: int = 100,
                 parent: QtWidgets.QWidget | None = None,
                 ) -> None:
        pass
```

**OPTION 2: Black style**

Black mangles this like so:

```python
    @typing.overload
    def __init__(
        self,
        figsize: tuple[float, float] = (5.0, 4.0),
        dpi: int = 100,
        parent: QtWidgets.QWidget | None = None,
    ) -> None:
        pass
```

**OPTION 3:  Single line (hard to read)**

Could wrap this at 80 (or 100) chars or something... but it remains hard to read.

```python
    @typing.overload
    def __init__(self, figsize: tuple[float, float] = (5.0, 4.0), dpi: int = 100, parent: QtWidgets.QWidget | None = None,) -> None:
        pass
```

**Comment by j9ac9k** [comment]
On mobile but I'm partial to black (and perhaps ruff?) style.

**Comment by NeilGirdhar** [comment]
Good idea asking this.  I would definitely go with what the project owner prefers.  One advantage is that you will be able to use Ruff or Black to automatically format.

(Just for the record though, option 1 is my personal style.)

**Comment by j9ac9k** [comment]
Honestly for this sort of thing I like the idea that individual
contributors can pass their own style into the library. So if a
contribution came with the first option I wouldn’t ask them to change it.
The word wrapped line  (option 3) is a no go haha

**Comment by rwarren** [comment]
I think we have a happy medium, then. :)

One of the great things about ruff is that it deviates slightly from black on the formatting in this respect.  The ruff formatter is perfectly fine with option 1 when it exists, while black wrecks it (personal opinion) with the dedenting of args.

I will stick with Option 1 for cases where I do it.  Realistically, I likely won't get all formatting nice and pretty since the focus will be on types, but... we shall see.

One other comparatively mild hiccup I've just noticed is that the pyqtgraph code-base has a lot of random trailing whitespace in files.  My editing is automatically stripping trailing whitespace.  Although this is non-typing noise in the diffs... **does anyone object to trailing whitespace being removed as part of this**?  Code impact is nil.

**Comment by j9ac9k** [comment]
I could use some input from those of you more versed with the `typing` module than I am (which is probably everyone here).

I'm working on redoing the docs for `PlotDataItem.py`  `PlotDataItem.setData()` accepts `*args, **kwargs` with a _mountain_ of possible `kwargs`.  

I'm considering doing something like

```python
class PointStyleKeywordArgs(TypedDict):
    symbol: str
    symbolPen: fn.color_like | QtGui.QPen
    symbolBrush: fn.color_like | QtGui.QPen | QtGui.QBrush
    symbolSize: int | Iterable[int]
    pxMode: bool

class LineStyleKeywordArgs(TypedDict):
    connect: str | np.ndarray
    pen: fn.color_like | QtGui.QPen | None
    shadowPen: fn.color_like | QtGui.QPen | None
    fillLevel: bool | None
    fillOutline: bool
    fillBrush: fn.color_like | QtGui.QBrush | None
    stepMode: str | None

class OptimizationKeywordArgs(TypedDict):
    useCache: bool
    antialias: bool
    downsample: int
    downsampleMethod: str
    autoDownsample: bool
    clipToView: bool
    dynamicRangeLimit: float | None
    dynamicRangeHyst: float
    skipFiniteCheck: bool
...

def setData(*args, **kwargs: PointStyleKeywordArgs | OptimizationKeywordArgs | LineStyleKeywordArgs):
   ...
```

Is there some better way I should be looking at doing the type annotations there?  I was considering specifying every possible accepted keyword argument in the function signature, but in hindsight, I don't think that's a great idea.

Suggestions would be greatly appreciated!
